align/compiler/gen_abstract_name.py

```diff
@@ -0,0 +1,141 @@
+            str: unique hex key
+        """
+        skeys = sorted(param.keys())
+        arg_str = '_'.join([k+':'+str(param[k]) for k in skeys])
+        key = f"_{str(int(hashlib.sha256(arg_str.encode('utf-8')).hexdigest(), 16) % 10**8)}"
+        return key
+
+    def group_cap_subcircuit(self, unit_cap):
+        #TODO hack for group cap, need to be fixed
+        """create subckt corresponding to unit cap
+        Args:
+            name (str): unique cap name in constraint
+        """
+        if not self.plib.find(unit_cap):
+            logger.debug(f"creating subcircuit for {unit_cap}")
+            cmodel = self.plib.find('CAP')
+            assert cmodel, f"no cap model found for groupcap constraint {cmodel}"
+            unit_cap_value = float(unit_cap.split('_')[1].replace('F', ''))*10E-15
+
+            with set_context(self.plib):
+                new_subckt = SubCircuit(name=unit_cap, pins=list(cmodel.pins), generator={"name": 'CAP'})
+            with set_context(new_subckt.elements):
+                new_ele = Instance(name='C0', model='CAP',
+                                   pins={pin: pin for pin in cmodel.pins},
+                                   parameters={'VALUE': unit_cap_value}
+                                   )
+                new_subckt.elements.append(new_ele)
+            self.plib.append(new_subckt)
+
+    def create_subckt(self, element, name):
+        """create_subckt
+        Adds a subckt in primitive library for a generic device instance if not already existing
+        Args:
+            element (instance): instance in a subcircuit
+            name (str): unique name for this instance based on parameters
+        """
+        if not self.plib.find(name):
+            logger.debug(f"creating subcircuit for {element}")
+            with set_context(self.plib):
+                new_subckt = SubCircuit(name=name, pins=list(element.pins.keys()), generator={"name":element.model})
+            with set_context(new_subckt.elements):
+                new_ele = Instance(name=element.name,
+                                   model=element.model,
+                                   pins={x: x for x in element.pins.keys()},
+                                   parameters=element.parameters
+                                   )
+                new_subckt.elements.append(new_ele)
+            self.plib.append(new_subckt)
+
+    def gen_primitive_def(self, element):
+        """gen_primitive_def
+
+            Adds subcircuits to primitive library for each instance with a different parameter
+
+        Args:
+            element (instance): instance properties
+        """
+        model = element.model
+        generator = self.ckt_lib.find(model)
+
+        if isinstance(generator, SubCircuit):
+            element.add_abs_name(model)
+            gen_const = [True for const in generator.constraints if isinstance(const, constraint.Generator)]
+            if gen_const and not self.plib.find(generator.name):
```

Comment:
@kkunal1408 , is this safe to do? This basically means that annotation is creating subcircuits that are duplicate or collision, right?

Reply:
"not self.plib.find(generator.name)" in line 126 checks for existing subcircuit to avoid duplicate